pebbles/tests/test_flask_api.py

```diff
@@ -2151,6 +2151,106 @@ def test_user_fetch_instance_usage_stats(self):
+        ns_kv.value = {'attr': 'val'}
+        ts = 1000000000
+        ns_kv.created_ts = ts
+        ns_kv.updated_ts = ts
+        db.session.add(ns_kv)
+        db.session.commit()
+
+        response = self.make_authenticated_admin_request(
+            method='PUT',
+            path='/api/v1/namespaced_keyvalues/%s/%s' % (ns, key),
+            data=json.dumps({'namespace': 'TestDriver', 'key': 'test_pool_1', 'value': {"attr": "val"}, 'updated_version_ts': ts})
+        )
+        self.assert_200(response)
+
+    def test_invalid_create_update_namespaced_data(self):
+        response = self.make_authenticated_admin_request(
+            method='POST',
+            path='/api/v1/namespaced_keyvalues',
+            data=json.dumps({'namespace': 'TestDriver', 'key': 0, 'value': {"attr": "val"}})
+        )
+        self.assertStatus(response, 422)
+
+        response = self.make_authenticated_admin_request(
+            method='POST',
+            path='/api/v1/namespaced_keyvalues',
+            data=json.dumps({'namespace': '', 'key': 'test_pool_vm_4', 'value': {"attr": "val"}})
+        )
+        self.assertStatus(response, 422)
+
+        ns = "TestDriver"
+        key = "pool_vm_test_4"
+        ns_kv = NamespacedKeyValue(ns, key)
+        ns_kv.value = {'attr': 'val'}
+        ts = 1000000000
+        ns_kv.created_ts = ts
+        ns_kv.updated_ts = ts
+        db.session.add(ns_kv)
+        db.session.commit()
+
+        response = self.make_authenticated_admin_request(
+            method='PUT',
+            path='/api/v1/namespaced_keyvalues/%s/%s' % (ns, key),
+            data=json.dumps({'namespace': 'TestDriver', 'key': 'test_pool_1', 'value': {"attr": "val"}, 'updated_version_ts': 'bogusx10'})
+        )
+        self.assertStatus(response, 422)
+
+    def test_concurrent_update_namespaced_data(self):
+        namespace = "TestDriver"
+        key = "test_pool_vm_5"
+        namespace_keyvalue_obj = NamespacedKeyValue(namespace, key)
+        namespace_keyvalue_obj.value = {'attribute': 'value'}
+        ts = 1000000000
+        namespace_keyvalue_obj.created_ts = ts
+        namespace_keyvalue_obj.updated_ts = ts
+        db.session.add(namespace_keyvalue_obj)
+        db.session.commit()
+
+        response = self.make_authenticated_admin_request(
+            method='PUT',
+            path='/api/v1/namespaced_keyvalues/%s/%s' % (namespace, key),
+            data=json.dumps({'namespace': namespace, 'key': key, 'value': {"attr": "val"}, 'updated_version_ts': ts})
+        )
+        self.assert_200(response)
+
+        # Concurrent Modification
+        invalid_response = self.make_authenticated_admin_request(
+            method='PUT',
+            path='/api/v1/namespaced_keyvalues/%s/%s' % (namespace, key),
+            data=json.dumps({'namespace': namespace, 'key': key, 'value': {"another_attr": "another_val"}, 'updated_version_ts': ts})
+        )
+        self.assertStatus(invalid_response, 409)
```

Comment:
Good point, but they are practically everywhere in the tests, cannot change them all in this commit at least, can do as a part of general re-factoring later